reactos/drivers/usb/usbhub_new/pnp.c

```diff
@@ -0,0 +1,2822 @@
+                  IN BOOLEAN IsEnable)
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+    PVOID NameBuffer;
+
+    DPRINT("USBH_SymbolicLink ... \n");
+
+    if (IsEnable)
+    {
+        Status = IoRegisterDeviceInterface(PortExtension->Common.SelfDevice,
+                                           InterfaceClassGuid,
+                                           NULL,
+                                           &PortExtension->SymbolicLinkName);
+
+        if (NT_SUCCESS(Status))
+        {
+            USBH_SetPdoRegistryParameter(PortExtension->Common.SelfDevice,
+                                         L"SymbolicName",
+                                         PortExtension->SymbolicLinkName.Buffer,
+                                         PortExtension->SymbolicLinkName.Length,
+                                         REG_SZ,
+                                         PLUGPLAY_REGKEY_DEVICE);
+
+            Status = IoSetDeviceInterfaceState(&PortExtension->SymbolicLinkName,
+                                               TRUE);
+        }
+    }
+    else
+    {
+        NameBuffer = PortExtension->SymbolicLinkName.Buffer;
+
+        if (NameBuffer)
+        {
+            Status = IoSetDeviceInterfaceState(&PortExtension->SymbolicLinkName,
+                                               FALSE);
+
+            ExFreePool(PortExtension->SymbolicLinkName.Buffer);
+
+            PortExtension->SymbolicLinkName.Buffer = NULL;
+        }
+    }
+
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+USBH_RestoreDevice(IN PUSBHUB_PORT_PDO_EXTENSION PortExtension,
+                   IN BOOLEAN IsKeepDeviceData)
+{
+    PUSBHUB_FDO_EXTENSION HubExtension;
+    PUSBHUB_PORT_DATA PortData;
+    NTSTATUS Status;
+    ULONG ix;
+
+    DPRINT("USBH_RestoreDevice ... \n");
+
+    HubExtension = PortExtension->HubExtension;
+
+    if (!HubExtension)
+    {
+        Status = STATUS_UNSUCCESSFUL;
+        return Status;
+    }
+
+    ASSERT(PortExtension->PortNumber > 0);
+    PortData = &HubExtension->PortData[PortExtension->PortNumber - 1];
+
+    if (PortExtension->Common.SelfDevice == PortData->DeviceObject)
+    {
+        Status = STATUS_UNSUCCESSFUL;
+        return Status;
+    }
+
+    Status = USBH_SyncGetPortStatus(HubExtension,
+                                    PortExtension->PortNumber,
+                                    &PortData->PortStatus,
+                                    sizeof(USBHUB_PORT_STATUS));
+
+    if (NT_SUCCESS(Status))
+    {
+        for (ix = 0; ix < 3; ix++)
+        {
+            Status = USBH_ResetDevice((PUSBHUB_FDO_EXTENSION)HubExtension,
+                                      PortExtension->PortNumber,
+                                      IsKeepDeviceData,
+                                      ix == 0);
+
+            if (NT_SUCCESS(Status) || Status == STATUS_NO_SUCH_DEVICE)
+            {
+                break;
+            }
+
+            USBH_Wait(1000);
+      
```

Comment:
pnp.c:2186: Can HubExtension really be NULL here? If yes, you are missing null checks below, e.g. L2264 dereferences it without a check.